pi.py

```diff
@@ -1 +1,578 @@
+                TOB=np.nan
+                LNB=np.nan
+                IFLAG=3
+                # Return the unsuitable values
+                return(CAPED,TOB,LNB,IFLAG)
+            else:
+                first_lvl=first_valid
+    else:
+        first_lvl=0
+
+    # Populate new environmental profiles removing values above ptop and
+    # find new number, N, of profile levels with which to calculate CAPE
+    N=np.argmin(np.abs(P-ptop))
+    
+    P=P[first_lvl:N]
+    T=T[first_lvl:N]
+    R=R[first_lvl:N]
+    nlvl=len(P)
+    TVRDIF = np.zeros((nlvl,))
+    
+    #
+    #   ***  Run checks   ***
+    #
+
+    # CHECK: Is the input parcel suitable? If not, return missing CAPE
+    if ((RP < 1e-6) or (TP < 200)):
+        CAPED=0
+        TOB=np.nan
+        LNB=np.nan
+        IFLAG=0
+        # Return the unsuitable values
+        return(CAPED,TOB,LNB,IFLAG)
+    
+    #
+    #   ***  Define constants   ***
+    #
+
+    # Thermodynamic Constants
+    CPD=1005.7       # [J/kg.K] Specific heat of dry air at constant pressure
+    CPV=1870.0       # [J/kg.K] Specific heat of water vapor at constant pressure
+    #CL=4190.0       # [J/kg.K] Specific heat of liquid water
+    CL=2500.0        # [J/kg.K] Modified specific heat of liquid water
+    CPVMCL=CPV-CL
+    RV=461.5         # [J/kg.K] gas constant of water vapor
+    RD=287.04        # [J/kg.K] gas constant of dry air
+    EPS=RD/RV        # [unitless] epsilon, the ratio of gas constants
+    ALV0=2.501e6     # [J/kg] latent heat of vaporization at 0 degrees C
+    
+    #
+    #  ***  Define various parcel quantities, including reversible   ***
+    #  ***                       entropy, S                          ***
+    #                         
+    TPC=TP-273.15                           # Parcel temperature in Celsius
+    ESP=6.112*np.exp(17.67*TPC/(243.5+TPC)) # Parcel's saturated vapor pressure
+    EVP=RP*PP/(EPS+RP)                      # Parcel's partial vapor pressure
+    RH=EVP/ESP                              # Parcel's relative humidity
+    RH=min([RH,1.0])                     # ensure that the relatively humidity does not exceed 1.0
+    ALV=ALV0+CPVMCL*TPC                     # calculate the latent heat of vaporization, dependant on temperature
+    # calculate reversible total specific entropy per unit mass of dry air (E94, EQN. 4.5.9)
+    S=(CPD+RP*CL)*np.log(TP)-RD*np.log(PP-EVP)+ALV*RP/TP-RP*RV*np.log(RH)
+    
+    #
+    #   ***  Estimate lifted condensation level pressure, PLCL   ***
+    #     Based on E94 "calcsound.f" code at http://texmex.mit.edu/pub/emanuel/BOOK/
+    #     see also https://psl.noaa.gov/data/composites/day/calculation.html
+    #
+    #   NOTE: Modern PLCL calculations are made following the exact expressions of Romps (2017),
+    #   see https://journals.ametsoc.org/doi/pdf/10.1175/JAS-D-17-0102.1
+    #   and Python PLCL code at http://romps.berkeley.edu/papers/pubdata/2016/lcl/lcl.py
+    #
+    CHI=TP/(1669.0-122.0*RH-TP)
```

Comment:
Can you define these constants? 1669 and 122?

Reply:
They are discussed in the documentation, just labeled A and B. Kerry developed this empirical way of finding p_LCL but it isn't documented anywhere except this code and his convection code for his textbook. As noted in the Projects of the repository, my long term goal is to use David Romps' formulation eventually.

I can add A/B to the constants.py file.